### PR TITLE
feat: PC-12012 Composite SLO validation

### PR DESCRIPTION
**Pull request by marcinlawnik**
## Motivation

Added more validation and tests according to specification

## Summary

Validation for `kind.SLO`

### CODE REVIEW

File: manifest/v1alpha/slo/validation.go [hunk 1]
```diff
@@ -37,27 +44,96 @@ var specValidationNonComposite = validation.New[Spec](
...
+		),
+
+	validation.For(func(s SLO) []CompositeObjective { return s.Spec.Objectives[0].Composite.Objectives }).
+		StopOnError().
+		Rules(
+			validation.NewSingleRule(func(c []CompositeObjective) error {
+				sloMap := make(map[string]bool)
+
+				for objKey, obj := range c {
+					key := fmt.Sprintf("%s%s%s", obj.Project, obj.SLO, obj.Objective)
```
Comment (ditrytus) [inline manifest/v1alpha/slo/validation.go:118]: Assuming we have the following two objectives:
```
- project: a
  slo: bc
  objective: d
- project: a
  slo: b
  objective: cd
```
They will concatenate into identical key `abcd` and trigger a validation error. Let's add a separator that is forbidden in project, slo, objective name:
```suggestion
					key := fmt.Sprintf("%s/%s/%s", obj.Project, obj.SLO, obj.Objective)
```

File: manifest/v1alpha/slo/validation.go [hunk 2]
```diff
@@ -184,6 +260,13 @@ var specCompositeObjectiveValidationRule = validation.NewSingleRule(func(s Spec)
 	return nil
 })
 
+var compositeObjectiveRule = validation.New[CompositeObjective](
+	validation.ForPointer(func(c CompositeObjective) *float64 { return c.Weight }).
+		WithName("weight").
+		Required().
+		Rules(validation.GreaterThan(0.0)),
```
Comment (ditrytus) [inline manifest/v1alpha/slo/validation.go:267]: Fields `project`, `slo`, `objective` and `whenDelayed` are required. Additionally, we could validate if `project`, `slo`, `objective` are a proper subdomains, using `validation.StringIsDNSSubdomain()`. `project` and `name` are originally validated with `validationV1Alpha.FieldRuleMetadataName` and `validationV1Alpha.FieldRuleMetadataProject` but I am not sure if we should reuse those.

Comment (ditrytus) [inline manifest/v1alpha/slo/validation.go:267]: `whenDelayed` is an enum so it should be validated against allowed set of values.

**Review by ditrytus — CHANGES_REQUESTED** [review]

File: manifest/v1alpha/slo/validation.go [hunk 3]
```diff
@@ -10,6 +13,10 @@ import (
 )
 
 var sloValidation = validation.New[SLO](
+	validation.For(func(s SLO) SLO { return s }).
+		StopOnError().
+		Include(sloValidationComposite),
```
[the remark this answers is not in the document]
Comment (marcinlawnik) [reply manifest/v1alpha/slo/validation.go:18]: Changed

File: manifest/v1alpha/slo/validation.go [hunk 4]
```diff
@@ -37,27 +44,96 @@ var specValidationNonComposite = validation.New[Spec](
 var specValidationComposite = validation.New[Spec](
 	validation.For(func(s Spec) Indicator { return s.Indicator }).
 		WithName("indicator").
+		StopOnError().
```
[the remark this answers is not in the document]
Comment (marcinlawnik) [reply manifest/v1alpha/slo/validation.go:47]: Changed

File: manifest/v1alpha/slo/validation.go (hunk 1 above)
Comment (marcinlawnik) [reply manifest/v1alpha/slo/validation.go:118]: Makes sense, changed

File: manifest/v1alpha/slo/validation.go (hunk 2 above)
Comment (marcinlawnik) [reply manifest/v1alpha/slo/validation.go:267]: Added, with tests

Comment (marcinlawnik) [reply manifest/v1alpha/slo/validation.go:267]: Added, with test

File: manifest/v1alpha/slo/validation_composite_test.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+			err := validate(slo)
+			testutils.AssertContainsErrors(t, slo, err, 1, testutils.ExpectedError{
+				Prop:    "spec.indicator",
+				Code:    validation.ErrorCodeForbidden,
+				Message: "property is forbidden; indicator section is forbidden when spec.objectives[0].composite is provided",
+			})
+		}
+	})
+	t.Run("fails - other objective types mixed with composite", func(t *testing.T) {
+		for _, obj := range []Objective{
```
Comment (ditrytus) [inline manifest/v1alpha/slo/validation_composite_test.go:50]: Loop is used for a single element. We could add another one to test `CountMetricsSpec`.

**Review by ditrytus — APPROVED** [review]
One idea to add to the tests but otherwise good to go.

**Pull request merged**